File: BetterTridents/src/main/java/project/kompass/btk/listener/PotionSoupStackListener.java
package project.kompass.btk.listener;

import project.kompass.btk.BTK;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Item;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityPickupItemEvent;
import org.bukkit.event.entity.ItemSpawnEvent;
import org.bukkit.event.inventory.BrewEvent;
import org.bukkit.event.inventory.CraftItemEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCreativeEvent;
import org.bukkit.event.inventory.InventoryDragEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.java.JavaPlugin;

public class PotionSoupStackListener implements Listener {

    // Main optimization: checking materials first bypasses expensive ItemMeta allocations for 95%+ of items
    private void updateStackSize(ItemStack item) {
        if (item == null || item.getType() == Material.AIR) return;
        Material type = item.getType();

        // 1. Potions: Custom stack size 16
        if (type == Material.POTION || type == Material.SPLASH_POTION || type == Material.LINGERING_POTION) {
            ItemMeta meta = item.getItemMeta();
            if (meta != null && (!meta.hasMaxStackSize() || meta.getMaxStackSize() != 16)) {
                meta.setMaxStackSize(16);
                item.setItemMeta(meta);
            }
        }
        // 2. Soups and Stews: Custom stack size 8
        else if (type == Material.MUSHROOM_STEW || type == Material.RABBIT_STEW ||
                type == Material.BEETROOT_SOUP || type == Material.SUSPICIOUS_STEW) {
            ItemMeta meta = item.getItemMeta();
            if (meta != null && (!meta.hasMaxStackSize() || meta.getMaxStackSize() != 8)) {
                meta.setMaxStackSize(8);
                item.setItemMeta(meta);
            }
        }
    }

    private void updateInventory(Inventory inventory) {
        if (inventory == null) return;
        for (ItemStack item : inventory.getContents()) {
            updateStackSize(item);
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerJoin(PlayerJoinEvent event) {
        updateInventory(event.getPlayer().getInventory());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onInventoryOpen(InventoryOpenEvent event) {
        updateInventory(event.getInventory());
        updateInventory(event.getPlayer().getInventory());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onInventoryClick(InventoryClickEvent event) {
        // Skip creative clicks here; they are handled separately by onInventoryCreative
        if (event instanceof InventoryCreativeEvent) return;

        updateStackSize(event.getCurrentItem());
        updateStackSize(event.getCursor());

        Bukkit.getScheduler().runTask(JavaPlugin.getPlugin(BTK.class), () -> {
            if (event.getClickedInventory() != null) {
                updateStackSize(event.getClickedInventory().getItem(event.getSlot()));
            }
            updateStackSize(event.getCursor());
        });
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onInventoryCreative(InventoryCreativeEvent event) {
        updateStackSize(event.getCurrentItem());
        updateStackSize(event.getCursor());

        // Next-tick execution forces the server to send an update packet (SetSlot) back to
        // the client, overriding the client's local creative inventory assumptions.
        Bukkit.getScheduler().runTask(JavaPlugin.getPlugin(BTK.class), () -> {
            updateStackSize(event.getWhoClicked().getItemOnCursor());
            if (event.getClickedInventory() != null) {
                updateStackSize(event.getClickedInventory().getItem(event.getSlot()));
            }
        });
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onInventoryDrag(InventoryDragEvent event) {
        for (ItemStack item : event.getNewItems().values()) {
            updateStackSize(item);
        }
        updateStackSize(event.getOldCursor());
        updateStackSize(event.getCursor());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onItemSpawn(ItemSpawnEvent event) {
        Item itemEntity = event.getEntity();
        ItemStack item = itemEntity.getItemStack();
        updateStackSize(item);
        itemEntity.setItemStack(item);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onItemPickup(EntityPickupItemEvent event) {
        ItemStack item = event.getItem().getItemStack();
        updateStackSize(item);
        event.getItem().setItemStack(item);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBrew(BrewEvent event) {
        updateInventory(event.getContents());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onCraft(CraftItemEvent event) {
        updateStackSize(event.getCurrentItem());
        updateStackSize(event.getCursor());
        updateStackSize(event.getRecipe().getResult());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerInteract(PlayerInteractEvent event) {
        updateStackSize(event.getItem());
    }
}

File: BetterTridents/src/main/kotlin/project/kompass/btk/listener/PotionSoupStackListener.kt
package project.kompass.btk.listener

import project.kompass.btk.BTK
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityPickupItemEvent
import org.bukkit.event.entity.ItemSpawnEvent
import org.bukkit.event.inventory.*
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.java.JavaPlugin
import java.util.EnumSet

class PotionSoupStackListener : Listener {

    companion object {
        private val STACK_TARGET_MATERIALS: Set<Material> = EnumSet.of(
            Material.POTION,
            Material.SPLASH_POTION,
            Material.LINGERING_POTION,
            Material.MUSHROOM_STEW,
            Material.RABBIT_STEW,
            Material.BEETROOT_SOUP,
            Material.SUSPICIOUS_STEW,
            Material.CAKE,
            Material.TOTEM_OF_UNDYING
        )
    }

    // Nanosecond O(1) bitmask check
    private fun isStackTarget(item: ItemStack?): Boolean {
        return item != null && STACK_TARGET_MATERIALS.contains(item.type)
    }

    private fun updateStackSize(item: ItemStack?) {
        if (item == null || item.type == Material.AIR) return
        val type = item.type

        // 1. Potions: Custom stack size 8
        if (type == Material.POTION || type == Material.SPLASH_POTION || type == Material.LINGERING_POTION) {
            val meta = item.itemMeta
            if (meta != null && (!meta.hasMaxStackSize() || meta.maxStackSize != 8)) {
                meta.setMaxStackSize(8)
                item.itemMeta = meta
            }
        }
        // 2. Soups and Stews: Custom stack size 16
        else if (type == Material.MUSHROOM_STEW || type == Material.RABBIT_STEW ||
            type == Material.BEETROOT_SOUP || type == Material.SUSPICIOUS_STEW) {
            val meta = item.itemMeta
            if (meta != null && (!meta.hasMaxStackSize() || meta.maxStackSize != 16)) {
                meta.setMaxStackSize(16)
                item.itemMeta = meta
            }
        }
        // 3. Cakes: Custom stack size 16
        else if (type == Material.CAKE) {
            val meta = item.itemMeta
            if (meta != null && (!meta.hasMaxStackSize() || meta.maxStackSize != 16)) {
                meta.setMaxStackSize(16)
                item.itemMeta = meta
            }
        }
        // 4. Totems of Undying: Custom stack size 4
        else if (type == Material.TOTEM_OF_UNDYING) {
            val meta = item.itemMeta
            if (meta != null && (!meta.hasMaxStackSize() || meta.maxStackSize != 4)) {
                meta.setMaxStackSize(4)
                item.itemMeta = meta
            }
        }
    }

    private fun updateInventory(inventory: Inventory?) {
        if (inventory == null) return
        for (item in inventory.contents) {
            if (isStackTarget(item)) {
                updateStackSize(item)
            }
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        updateInventory(event.player.inventory)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onInventoryOpen(event: InventoryOpenEvent) {
        updateInventory(event.inventory)
        updateInventory(event.player.inventory)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onInventoryClick(event: InventoryClickEvent) {
        if (event is InventoryCreativeEvent) return

        val cursor = event.cursor
        val currentItem = event.currentItem

        if (!isStackTarget(cursor) && !isStackTarget(currentItem)) return

        updateStackSize(currentItem)
        updateStackSize(cursor)

        Bukkit.getScheduler().runTask(JavaPlugin.getPlugin(BTK::class.java), Runnable {
            event.clickedInventory?.let {
                val slotItem = it.getItem(event.slot)
                if (isStackTarget(slotItem)) {
                    updateStackSize(slotItem)
                    it.setItem(event.slot, slotItem)
                }
            }
            val player = event.whoClicked as? Player
            if (player != null && player.isOnline) {
                val cursorItem = player.itemOnCursor
                if (isStackTarget(cursorItem)) {
                    updateStackSize(cursorItem)
                    player.setItemOnCursor(cursorItem)
                }
            }
        })
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onInventoryCreative(event: InventoryCreativeEvent) {
        val cursor = event.cursor
        val currentItem = event.currentItem

        if (!isStackTarget(cursor) && !isStackTarget(currentItem)) return

        val oldCursor = event.whoClicked.itemOnCursor
        if (oldCursor != null && oldCursor.type != Material.AIR &&
            cursor != null && cursor.type != Material.AIR &&
            oldCursor.isSimilar(cursor)) {

            val type = oldCursor.type
            val isPotion = (type == Material.POTION || type == Material.SPLASH_POTION || type == Material.LINGERING_POTION)
            val isSoup = (type == Material.MUSHROOM_STEW || type == Material.RABBIT_STEW ||
                    type == Material.BEETROOT_SOUP || type == Material.SUSPICIOUS_STEW)
            val isCake = (type == Material.CAKE)
            val isTotem = (type == Material.TOTEM_OF_UNDYING)

            if (isPotion || isSoup || isCake || isTotem) {
                val max = when {
                    isPotion -> 8
                    isSoup -> 16
                    isCake -> 16
                    isTotem -> 4
                    else -> 64
                }
                val newAmount = Math.min(oldCursor.amount + cursor.amount, max)

                val stacked = oldCursor.clone()
                stacked.amount = newAmount
                updateStackSize(stacked)

                event.whoClicked.setItemOnCursor(stacked)
                event.setResult(org.bukkit.event.Event.Result.ALLOW)
            }
        } else {
            updateStackSize(currentItem)
            updateStackSize(cursor)
        }

        Bukkit.getScheduler().runTask(JavaPlugin.getPlugin(BTK::class.java), Runnable {
            val player = event.whoClicked as? Player ?: return@Runnable
            if (player.isOnline) {
                val cursorItem = player.itemOnCursor
                if (isStackTarget(cursorItem)) {
                    updateStackSize(cursorItem)
                    player.setItemOnCursor(cursorItem)
                }

                event.clickedInventory?.let {
                    val slotItem = it.getItem(event.slot)
                    if (isStackTarget(slotItem)) {
                        updateStackSize(slotItem)
                        it.setItem(event.slot, slotItem)
                    }
                }
            }
        })
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onInventoryDrag(event: InventoryDragEvent) {
        val oldCursor = event.oldCursor
        val cursor = event.cursor

        var involved = isStackTarget(oldCursor) || isStackTarget(cursor)
        if (!involved) {
            for (item in event.newItems.values) {
                if (isStackTarget(item)) {
                    involved = true
                    break
                }
            }
        }
        if (!involved) return

        for (item in event.newItems.values) {
            updateStackSize(item)
        }
        updateStackSize(oldCursor)
        updateStackSize(cursor)

        Bukkit.getScheduler().runTask(JavaPlugin.getPlugin(BTK::class.java), Runnable {
            for (slot in event.newItems.keys) {
                val slotItem = event.view.getItem(slot)
                if (isStackTarget(slotItem)) {
                    updateStackSize(slotItem)
                    event.view.setItem(slot, slotItem)
                }
            }
            val player = event.whoClicked as? Player
            if (player != null && player.isOnline) {
                val cursorItem = player.itemOnCursor
                if (isStackTarget(cursorItem)) {
                    updateStackSize(cursorItem)
                    player.setItemOnCursor(cursorItem)
                }
            }
        })
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onItemSpawn(event: ItemSpawnEvent) {
        val itemEntity = event.entity
        val item = itemEntity.itemStack
        if (!isStackTarget(item)) return

        updateStackSize(item)
        itemEntity.itemStack = item
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onItemPickup(event: EntityPickupItemEvent) {
        val itemEntity = event.item
        val item = itemEntity.itemStack
        if (!isStackTarget(item)) return

        updateStackSize(item)
        itemEntity.itemStack = item
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onBrew(event: BrewEvent) {
        updateInventory(event.contents)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onCraft(event: CraftItemEvent) {
        val result = event.recipe.result
        if (!isStackTarget(result)) return

        val currentItem = event.currentItem
        val cursor = event.cursor

        updateStackSize(currentItem)
        updateStackSize(cursor)
        updateStackSize(result)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerInteract(event: PlayerInteractEvent) {
        if (!event.hasItem()) return
        val item = event.item
        if (!isStackTarget(item)) return

        updateStackSize(item)
        event.player.inventory.setItem(event.hand ?: org.bukkit.inventory.EquipmentSlot.HAND, item)
    }
}